File: python/sdfglib/mlir_frontend.py
"""
MLIR Frontend utilities for Python

This module provides Python wrappers for MLIR-style type conversions
and operation mappings to SDFG library nodes.
"""

from typing import List, Dict, Tuple
from enum import Enum


class MLIRScalarType(Enum):
    """MLIR scalar types that can be converted to SDFG types"""

    I1 = "i1"
    I8 = "i8"
    I16 = "i16"
    I32 = "i32"
    I64 = "i64"
    F16 = "f16"
    F32 = "f32"
    F64 = "f64"
    INDEX = "index"


class SDFGPrimitiveType(Enum):
    """SDFG primitive types"""

    BOOL = "bool"
    INT8 = "int8"
    INT16 = "int16"
    INT32 = "int32"
    INT64 = "int64"
    HALF = "half"
    FLOAT = "float"
    DOUBLE = "double"


# Mapping from MLIR types to SDFG types
MLIR_TO_SDFG_TYPE_MAP: Dict[str, SDFGPrimitiveType] = {
    "i1": SDFGPrimitiveType.BOOL,
    "i8": SDFGPrimitiveType.INT8,
    "i16": SDFGPrimitiveType.INT16,
    "i32": SDFGPrimitiveType.INT32,
    "i64": SDFGPrimitiveType.INT64,
    "f16": SDFGPrimitiveType.HALF,
    "f32": SDFGPrimitiveType.FLOAT,
    "f64": SDFGPrimitiveType.DOUBLE,
    "index": SDFGPrimitiveType.INT64,
}

# Elementwise operation mappings
ELEMENTWISE_OPS: Dict[str, str] = {
    # Binary operations
    "add": "Add",
    "sub": "Sub",
    "mul": "Mul",
    "div": "Div",
    "pow": "Pow",
    "minimum": "Minimum",
    "maximum": "Maximum",
    # Unary operations
    "abs": "Abs",
    "sqrt": "Sqrt",
    "exp": "Exp",
    "erf": "Erf",
    "sigmoid": "Sigmoid",
    "tanh": "Tanh",
    "relu": "Relu",
    "leaky_relu": "LeakyRelu",
    "elu": "Elu",
    "hard_sigmoid": "HardSigmoid",
    "cast": "Cast",
}

# Reduction operation mappings
REDUCE_OPS: Dict[str, str] = {
    "sum": "Sum",
    "mean": "Mean",
    "std": "Std",
    "max": "Max",
    "min": "Min",
    "softmax": "Softmax",
}


def convert_scalar_type(mlir_type: str) -> SDFGPrimitiveType:
    """
    Convert MLIR scalar type string to SDFG primitive type

    Args:
        mlir_type: MLIR type string (e.g., "f32", "i64")

    Returns:
        SDFG primitive type

    Raises:
        ValueError: If the MLIR type is not supported
    """
    if mlir_type not in MLIR_TO_SDFG_TYPE_MAP:
        raise ValueError(f"Unsupported MLIR scalar type: {mlir_type}")
    return MLIR_TO_SDFG_TYPE_MAP[mlir_type]


def convert_tensor_type(element_type: str, shape: List[int]) -> Tuple[SDFGPrimitiveType, List[int]]:
    """
    Convert MLIR tensor type to SDFG flat pointer type

    In SDFG, tensors are represented as flat pointers to scalars.
    The shape information is returned separately.

    Args:
        element_type: Element type string (e.g., "f32")
        shape: Tensor shape dimensions

    Returns:
        Tuple of (element primitive type, shape)

    Raises:
        ValueError: If the element type is not supported
    """
    sdfg_type = convert_scalar_type(element_type)
    return (sdfg_type, shape)


def get_elementwise_op_code(op_name: str) -> str:
    """
    Get SDFG library node code for elementwise operation

    Args:
        op_name: MLIR operation name

    Returns:
        SDFG library node code string

    Raises:
        ValueError: If the operation is not supported
    """
    if op_name not in ELEMENTWISE_OPS:
        raise ValueError(f"Unsupported elementwise operation: {op_name}")
    return ELEMENTWISE_OPS[op_name]


def get_reduce_op_code(op_name: str) -> str:
    """
    Get SDFG library node code for reduction operation

    Args:
        op_name: MLIR operation name

    Returns:
        SDFG library node code string

    Raises:
        ValueError: If the operation is not supported
    """
    if op_name not in REDUCE_OPS:
        raise ValueError(f"Unsupported reduce operation: {op_name}")
    return REDUCE_OPS[op_name]


def is_elementwise_unary(op_name: str) -> bool:
    """Check if operation is elementwise unary"""
    unary_ops = {
        "abs",
        "sqrt",
        "exp",
        "erf",
        "sigmoid",
        "tanh",
        "relu",
        "leaky_relu",
        "elu",
        "hard_sigmoid",
        "cast",
    }
    return op_name in unary_ops


def is_elementwise_binary(op_name: str) -> bool:
    """Check if operation is elementwise binary"""
    binary_ops = {"add", "sub", "mul", "div", "pow", "minimum", "maximum"}
    return op_name in binary_ops


def is_reduce_op(op_name: str) -> bool:
    """Check if operation is a reduction"""
    return op_name in REDUCE_OPS

File: src/frontend/mlir_frontend.cpp
/**
 * @file mlir_frontend.cpp
 * @brief Implementation of MLIR-style frontend for SDFG type conversions
 */

#include "sdfg/frontend/mlir_frontend.h"

#include <sstream>
#include <stdexcept>

namespace sdfg {
namespace frontend {

// MLIR type string to SDFG PrimitiveType mapping
const std::unordered_map<std::string, types::PrimitiveType> MLIRFrontend::mlir_type_map_ = {
    {"i1", types::PrimitiveType::Bool},
    {"i8", types::PrimitiveType::Int8},
    {"i16", types::PrimitiveType::Int16},
    {"i32", types::PrimitiveType::Int32},
    {"i64", types::PrimitiveType::Int64},
    {"f16", types::PrimitiveType::Half},
    {"f32", types::PrimitiveType::Float},
    {"f64", types::PrimitiveType::Double},
    {"index", types::PrimitiveType::Int64}, // MLIR index type maps to i64
};

// MLIR elementwise operation name to SDFG library node code mapping
const std::unordered_map<std::string, std::string> MLIRFrontend::elementwise_op_map_ = {
    // Binary operations
    {"add", "Add"},
    {"sub", "Sub"},
    {"mul", "Mul"},
    {"div", "Div"},
    {"pow", "Pow"},
    {"minimum", "Minimum"},
    {"maximum", "Maximum"},

    // Unary operations
    {"abs", "Abs"},
    {"sqrt", "Sqrt"},
    {"exp", "Exp"},
    {"erf", "Erf"},
    {"sigmoid", "Sigmoid"},
    {"tanh", "Tanh"},
    {"relu", "Relu"},
    {"leaky_relu", "LeakyRelu"},
    {"elu", "Elu"},
    {"hard_sigmoid", "HardSigmoid"},
    {"cast", "Cast"},
};

// MLIR reduction operation name to SDFG library node code mapping
const std::unordered_map<std::string, std::string> MLIRFrontend::reduce_op_map_ = {
    {"sum", "Sum"},
    {"mean", "Mean"},
    {"std", "Std"},
    {"max", "Max"},
    {"min", "Min"},
    {"softmax", "Softmax"},
};

types::Scalar MLIRFrontend::convert_scalar_type(const std::string& mlir_type_str) {
    auto it = mlir_type_map_.find(mlir_type_str);
    if (it == mlir_type_map_.end()) {
        std::ostringstream oss;
        oss << "Unsupported MLIR scalar type: " << mlir_type_str;
        throw std::invalid_argument(oss.str());
    }

    return types::Scalar(it->second);
}

types::Pointer MLIRFrontend::convert_tensor_type(const std::string& element_type_str, const std::vector<int64_t>& shape) {
    // Convert element type to SDFG scalar
    types::Scalar element_type = convert_scalar_type(element_type_str);

    // Create flat pointer to scalar (tensor is represented as 1D array)
    // Shape information is used separately by library nodes
    return types::Pointer(element_type);
}

std::vector<symbolic::Expression> MLIRFrontend::shape_to_symbolic(const std::vector<int64_t>& shape) {
    std::vector<symbolic::Expression> symbolic_shape;
    symbolic_shape.reserve(shape.size());

    for (int64_t dim : shape) {
        symbolic_shape.push_back(symbolic::integer(dim));
    }

    return symbolic_shape;
}

std::string MLIRFrontend::get_elementwise_op_code(const std::string& op_name) {
    auto it = elementwise_op_map_.find(op_name);
    if (it == elementwise_op_map_.end()) {
        std::ostringstream oss;
        oss << "Unsupported elementwise operation: " << op_name;
        throw std::invalid_argument(oss.str());
    }

    return it->second;
}

std::string MLIRFrontend::get_reduce_op_code(const std::string& op_name) {
    auto it = reduce_op_map_.find(op_name);
    if (it == reduce_op_map_.end()) {
        std::ostringstream oss;
        oss << "Unsupported reduce operation: " << op_name;
        throw std::invalid_argument(oss.str());
    }

    return it->second;
}

bool MLIRFrontend::is_elementwise_unary(const std::string& op_name) {
    static const std::vector<std::string> unary_ops = {
        "abs", "sqrt", "exp", "erf", "sigmoid", "tanh", "relu", "leaky_relu", "elu", "hard_sigmoid", "cast"
    };

    return std::find(unary_ops.begin(), unary_ops.end(), op_name) != unary_ops.end();
}

bool MLIRFrontend::is_elementwise_binary(const std::string& op_name) {
    static const std::vector<std::string> binary_ops = {"add", "sub", "mul", "div", "pow", "minimum", "maximum"};

    return std::find(binary_ops.begin(), binary_ops.end(), op_name) != binary_ops.end();
}

bool MLIRFrontend::is_reduce_op(const std::string& op_name) {
    return reduce_op_map_.find(op_name) != reduce_op_map_.end();
}

} // namespace frontend
} // namespace sdfg
